ceno_zkvm/src/bin/e2e.rs

```diff
@@ -131,14 +135,77 @@ fn main() {
     );
 
     let max_steps = args.max_steps.unwrap_or(usize::MAX);
-    run_e2e(
-        program,
-        platform,
-        args.stack_size,
-        args.heap_size,
-        hints,
-        max_steps,
+
+    type E = GoldilocksExt2;
+    type Pcs = Basefold<GoldilocksExt2, BasefoldRSParams>;
+
+    let (prover, verifier, zkvm_witness, pi, cycle_num, e2e_start, exit_code) =
+        run_e2e_gen_witness::<E, Pcs>(
+            program,
+            platform,
+            args.stack_size,
+            args.heap_size,
+            hints,
+            max_steps,
+        );
+
+    let timer = Instant::now();
+    let mut zkvm_proof = run_e2e_proof(prover, zkvm_witness, pi);
+    let proving_time = timer.elapsed().as_secs_f64();
+    let e2e_time = e2e_start.elapsed().as_secs_f64();
+    let witgen_time = e2e_time - proving_time;
+    println!(
+        "Proving finished.\n\
+\tProving time = {:.3}s, freq = {:.3}khz\n\
+\tWitgen  time = {:.3}s, freq = {:.3}khz\n\
+\tTotal   time = {:.3}s, freq = {:.3}khz\n\
+\tthread num: {}",
+        proving_time,
+        cycle_num as f64 / proving_time / 1000.0,
+        witgen_time,
+        cycle_num as f64 / witgen_time / 1000.0,
+        e2e_time,
+        cycle_num as f64 / e2e_time / 1000.0,
+        rayon::current_num_threads()
     );
+
+    run_e2e_verify(&verifier, zkvm_proof.clone(), exit_code, max_steps);
+
+    // do sanity check
+    let transcript = Transcript::new(b"riscv");
+    // change public input maliciously should cause verifier to reject proof
+    zkvm_proof.raw_pi[0] = vec![<E as ff_ext::ExtensionField>::BaseField::ONE];
```

Comment:
That should be easier to read:

```suggestion
    zkvm_proof.raw_pi[0] = vec![Goldilocks::ONE];
```

(You need to import Goldilocks.)

Reply:
Indeed it improve readibility, fixed in https://github.com/scroll-tech/ceno/pull/690/commits